lib/recorder.dart

```diff
@@ -0,0 +1,1191 @@
+  }
+
+  _calledStartMeasureFrame = true;
+
+  if (!profile.isWarmingUp) {
+    // Tell the browser to mark the beginning of the frame.
+    html.window.performance.mark('measured_frame_start#$_currentFrameNumber');
+
+    _isMeasuringFrame = true;
+  }
+}
+
+/// Signals the end of a measured frame.
+///
+/// See [startMeasureFrame] for details on what this instrumentation is used
+/// for.
+///
+/// Warm-up frames are not measured. If [profile.isWarmingUp] was true
+/// when the corresponding [startMeasureFrame] was called,
+/// this function does nothing.
+void endMeasureFrame() {
+  if (!_calledStartMeasureFrame) {
+    throw Exception('`startMeasureFrame` has not been called before calling `endMeasureFrame`');
+  }
+
+  _calledStartMeasureFrame = false;
+
+  if (_isMeasuringFrame) {
+    // Tell the browser to mark the end of the frame, and measure the duration.
+    html.window.performance.mark('measured_frame_end#$_currentFrameNumber');
+    html.window.performance.measure(
+      'measured_frame',
+      'measured_frame_start#$_currentFrameNumber',
+      'measured_frame_end#$_currentFrameNumber',
+    );
+
+    // Increment the current frame number.
+    _currentFrameNumber += 1;
+
+    _isMeasuringFrame = false;
+  }
+}
+
+/// A function that receives a benchmark value from the framework.
+typedef EngineBenchmarkValueListener = void Function(num value);
+
+// Maps from a value label name to a listener.
+final Map<String, EngineBenchmarkValueListener> _engineBenchmarkListeners = <String, EngineBenchmarkValueListener>{};
+
+/// Registers a [listener] for engine benchmark values labeled by [name].
+///
+/// If another listener is already registered, overrides it.
+void registerEngineBenchmarkValueListener(String name, EngineBenchmarkValueListener listener) {
+  if (listener == null) {
+    throw ArgumentError(
+      'Listener must not be null. To stop listening to engine benchmark values '
+      'under label "$name", call stopListeningToEngineBenchmarkValues(\'$name\').',
+    );
+  }
+
+  if (_engineBenchmarkListeners.containsKey(name)) {
+    throw StateError(
+      'A listener for "$name" is already registered.\n'
+      'Call `stopListeningToEngineBenchmarkValues` to unregister the previous '
+      'listener before registering a new one.'
+    );
+  }
+
+  if (_engineBenchmarkListeners.isEmpty) {
+    // The first listener is being registered. Register the global listener.
+    js_util.setProperty(html.window, '_flutter_internal_on_benchmark', _dispatchEngineBenchmarkValue);
+  }
+
+  _engineBenchmarkListeners[name] = listener;
+}
+
+/// Stops listening to engine benchmark values under labeled by [name].
+void stopListeningToEngineBenchmarkValues(String name) {
+  _engineBenchmarkListeners.remove(name);
+  if (_engineBenchmarkListeners.isEmpty) {
+    // The last listener unregistered. Remove the global listener.
+    js_util.setProperty(html.window, '_flutter_internal_on_benchmark', null);
```

Comment:
What would be the process for changing this logic when the internals of the engine need to change? Can this part move to the flutter/flutter repo so that this package is independent of engine-benchmark protocol?

Reply:
This part is required by `WidgetRecorder`.

If we should move it to `flutter/flutter`, where should we move it to? It must be imported by `web_benchmarks`.